_episodes/l4_04_parameterized_tests.md

```diff
@@ -0,0 +1,353 @@
+TestValues values[] = {
+    // value are in format: salary, basic_bonus, years_employed, tax
+    TestValues{8000, 2000, 3, 0},
+    TestValues{8000, 2000, 11, 100},
+    TestValues{15000, 3000, 4, 800},
+    TestValues{26000, 4000, 12, 3200},
+    TestValues{60000, 8000, 13, 16500}
+};
+
+// Test that the tax calculation is correct.
+TEST_P(EmployeeTestParameterisedFixture, TaxCalculationIsCorrect) {
+    TestValues current_test_case_value = GetParam();
+    employee.setBaseSalary(current_test_case_value.inp_salary);
+    employee.SetBasicBonus(current_test_case_value.inp_bonus);
+    employee.setNumberYearsEmployed(current_test_case_value.inp_years_employed);
+    EXPECT_EQ(employee.getTaxAmount(), current_test_case_value.out_tax);
+}
+
+// Instantiate the test case with the values array.
+INSTANTIATE_TEST_SUITE_P( CheckTaxCalculation, 
+                          EmployeeTestParameterisedFixture,
+                          ValuesIn(values));
+```
+
+The major change as compared to our previous example is shown in the cell below and this change is responsible to generate a parameterised test using a test fixture.
+
+```cpp
+class EmployeeTestParameterisedFixture : public EmployeeTestFixture, 
+                                         public WithParamInterface<TestValues> {
+};
+```
+
+In addition, we used a function `GetParam()` defined in `gtest.h`. This function can help us to get the input values passed via `ValuesIn()` function and use it in the test logic according to our requirements. In this case, it helps us to retrieve 4 values in the order `inp_salary`, `inp_bonus`, `inp_years_employed` and `out_tax` for each test case. Thus, `GetParam()` provides a convenient way to retrieve multiple values and use them in our test logic.
+
+## 4. Advantages of Parameterised tests
+
+From the above discussion, we can see that the parameterised tests have the following advantages.
+
+1. **Code Reusability**: With parameterized tests, we can write a single test case that can be executed with different inputs or test data. This promotes code reusability by eliminating the need to duplicate similar test cases. Instead, we can define the test logic once and apply it to multiple scenarios, reducing code duplication and improving maintainability.
+
+2. **Increased Test Coverage**: Parameterized tests allows us to easily test a wide range of input values or test cases without writing separate test cases for each variation. This enables us to achieve better test coverage by covering various combinations, edge cases, and boundary values in a concise manner.
+
+3. **Simplified Test Maintenance**: When changes are required in the test logic, having parameterized tests simplifies the maintenance process. Instead of modifying multiple test cases individually, we only need to update the single test case, which will automatically be executed with the new test data. This saves time and effort in maintaining and updating the tests.
```

Comment:
```suggestion
3. **Simplified Test Maintenance**: When changes are required in the test logic, having parameterised tests simplifies the maintenance process. Instead of modifying multiple test cases individually, we only need to update the single test case, which will automatically be executed with the new test data. This saves time and effort in maintaining and updating the tests.
```